### PR TITLE
update requirements

### PR DESCRIPTION
**Pull request by vincentsarago**
## Description

This is the first PR which will update the requirements for the different PC services. 

The test should fail right now, but I wanted to see if they were passing for `pccommon` 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

Please delete options that are not relevant.

- [ ] I have performed a self-review
- [ ] Changelog has been updated
- [ ] Documentation has been updated
- [ ] Unit tests pass locally (./scripts/test)
- [ ] Code is linted and styled (./scripts/format)

### CODE REVIEW

File: docker-compose.dev.yml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 services:
   stac-dev:
+    platform: linux/amd64
```
Comment (vincentsarago) [inline docker-compose.dev.yml:3]: 🤦 will remove

File: pccommon/pccommon/config/collections.py [hunk 2]
```diff
@@ -19,11 +19,13 @@ def __str__(self) -> str:
...
-        allow_population_by_field_name = True
-        json_loads = orjson.loads
-        json_dumps = orjson_dumps
+
+    model_config = {
+        # TODO, see if we can use pydantic native function
+        # https://docs.pydantic.dev/latest/api/config/#pydantic.alias_generators.to_camel
+        "alias_generator": camelize,
+        "populate_by_name": True,
+    }
```
Comment (vincentsarago) [inline pccommon/pccommon/config/collections.py:28]: TODO

File: pccommon/pccommon/config/containers.py [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
-    class Config:
-        json_loads = orjson.loads
-        json_dumps = orjson_dumps
+    # json_loads/json_dumps config have been removed
+    # the authors seems to indicate that parsing/serialization
+    # in Rust (pydantic-core) is fast (but maybe not as fast as orjson)
+    # https://github.com/pydantic/pydantic/discussions/6388
+    # class Config:
+    #     json_loads = orjson.loads
+    #     json_dumps = orjson_dumps
```
Comment (vincentsarago) [inline pccommon/pccommon/config/containers.py:17]: we can't customize this with pydantic 2 but it seems that pydantic 2 is faster anyway

File: pcstac/pcstac/client.py [hunk 4]
```diff
@@ -45,18 +45,6 @@ class PCClient(CoreCrudClient):
...
-        """Generate conformance classes list."""
-        base_conformance_classes = self.base_conformance_classes.copy()
-
-        for extension in self.extensions:
-            extension_classes = getattr(extension, "conformance_classes", [])
-            base_conformance_classes.extend(extension_classes)
-
-        base_conformance_classes.extend(self.extra_conformance_classes)
-
-        return sorted(list(set(base_conformance_classes)))
```
Comment (vincentsarago) [inline pcstac/pcstac/client.py:58]: this was the same as https://github.com/stac-utils/stac-fastapi/blob/80064f77739ddc1acaafa78cb0d8a11f77ecd33e/stac_fastapi/types/stac_fastapi/types/core.py#L348-L357

File: pcstac/pcstac/filter.py [hunk 5]
```diff
@@ -34,6 +33,4 @@ async def _fetch() -> Dict:
                 return queryables
 
         cache_key = f"{CACHE_KEY_QUERYABLES}:{collection_id}"
-        queryables = await cached_result(_fetch, cache_key, request)
-        headers = {"Content-Type": "application/schema+json"}
-        return JSONResponse(queryables, headers=headers)
```
Comment (vincentsarago) [inline pcstac/pcstac/filter.py:39]: should be handled by https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/extensions/stac_fastapi/extensions/core/filter/filter.py#L96-L110

File: pcstac/pcstac/main.py [hunk 6]
```diff
@@ -51,6 +55,16 @@
...
 search_post_request_model = create_post_request_model(EXTENSIONS, base_model=PCSearch)
 
+
+@asynccontextmanager
+async def lifespan(app: FastAPI) -> AsyncGenerator:
+    """FastAPI Lifespan."""
+    await connect_to_db(app)
+    await connect_to_redis(app)
+    yield
+    await close_db_connection(app)
```
Comment (vincentsarago) [inline pcstac/pcstac/main.py:65]: startup/shutdown events are deprecated and `lifespan` is the `new` way to do

File: pcstac/pcstac/main.py [hunk 7]
```diff
@@ -63,11 +77,28 @@
...
+        default_response_class=ORJSONResponse,
+        lifespan=lifespan,
+    ),
     search_get_request_model=search_get_request_model,
     search_post_request_model=search_post_request_model,
     response_class=ORJSONResponse,
     exceptions={**DEFAULT_STATUS_CODES, **PC_DEFAULT_STATUS_CODES},
+    middleware=[
+        Middleware(BrotliMiddleware),
+        Middleware(ProxyHeaderMiddleware),
```
Comment (vincentsarago) [inline pcstac/pcstac/main.py:91]: both are defined in the default stac-fastapi so they should have been present in the application before

**Comment by vincentsarago** [comment]
I don't seem to be able to run the tests locally: 
```
scripts/test --common
WARN[0000] The "APP_INSIGHTS_INSTRUMENTATION_KEY" variable is not set. Defaulting to a blank string. 
WARN[0000] The "APP_INSIGHTS_INSTRUMENTATION_KEY" variable is not set. Defaulting to a blank string. 
[+] Running 4/0
 ⠿ Container pc-stac-db                       Running                                                                                                                                                                                                                                                                                0.0s
 ⠿ Container planetary-computer-apis-redis-1  Running                                                                                                                                                                                                                                                                                0.0s
 ⠿ Container pcapis-azurite                   Running                                                                                                                                                                                                                                                                                0.0s
 ⠿ Container planetary-computer-apis-stac-1   Running                                                                                                                                                                                                                                                                                0.0s
Running mypy for common...
Success: no issues found in 29 source files
Running black for common...
All done! ✨ 🍰 ✨
29 files would be left unchanged.
Running flake8 for common...
Running unit tests for common...
========================================================================================================================================================== test session starts ===========================================================================================================================================================
platform linux -- Python 3.9.19, pytest-7.4.4, pluggy-1.5.0
rootdir: /opt/src
configfile: pytest.ini
plugins: asyncio-0.18.3, anyio-3.7.1
asyncio: mode=auto
collected 3 items / 3 errors                                                                                                                                                                                                                                                                                                             

================================================================================================================================================================= ERRORS =================================================================================================================================================================
____________________________________________________________________________________________________________________________________________ ERROR collecting pccommon/tests/test_timeouts.py ____________________________________________________________________________________________________________________________________________
ImportError while importing test module '/opt/src/pccommon/tests/test_timeouts.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pccommon/tests/test_timeouts.py:11: in <module>
    from pccommon.middleware import add_timeout
pccommon/pccommon/middleware.py:19: in <module>
    from pccommon.tracing import trace_request
pccommon/pccommon/tracing.py:14: in <module>
    from pccommon.config import get_apis_config
pccommon/pccommon/config/__init__.py:4: in <module>
    from pccommon.config.core import PCAPIsConfig
pccommon/pccommon/config/core.py:8: in <module>
    from pydantic_settings import BaseSettings
/usr/lib/python3.9/site-packages/pydantic_settings/__init__.py:1: in <module>
    from .main import BaseSettings, SettingsConfigDict
/usr/lib/python3.9/site-packages/pydantic_settings/main.py:7: in <module>
    from pydantic._internal._config import config_keys
E   ModuleNotFoundError: No module named 'pydantic._internal'
____________________________________________________________________________________________________________________________________________ ERROR collecting pccommon/tests/test_tracing.py _____________________________________________________________________________________________________________________________________________
ImportError while importing test module '/opt/src/pccommon/tests/test_tracing.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pccommon/tests/test_tracing.py:1: in <module>
    from pccommon.tracing import _parse_cqljson
pccommon/pccommon/tracing.py:14: in <module>
    from pccommon.config import get_apis_config
pccommon/pccommon/config/__init__.py:4: in <module>
    from pccommon.config.core import PCAPIsConfig
pccommon/pccommon/config/core.py:8: in <module>
    from pydantic_settings import BaseSettings
/usr/lib/python3.9/site-packages/pydantic_settings/__init__.py:1: in <module>
    from .main import BaseSettings, SettingsConfigDict
/usr/lib/python3.9/site-packages/pydantic_settings/main.py:7: in <module>
    from pydantic._internal._config import config_keys
E   ModuleNotFoundError: No module named 'pydantic._internal'
______________________________________________________________________________________________________________________________________ ERROR collecting pccommon/tests/config/test_render_config.py ______________________________________________________________________________________________________________________________________
ImportError while importing test module '/opt/src/pccommon/tests/config/test_render_config.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
pccommon/tests/config/test_render_config.py:3: in <module>
    from pccommon.config import get_render_config
pccommon/pccommon/config/__init__.py:4: in <module>
    from pccommon.config.core import PCAPIsConfig
pccommon/pccommon/config/core.py:8: in <module>
    from pydantic_settings import BaseSettings
/usr/lib/python3.9/site-packages/pydantic_settings/__init__.py:1: in <module>
    from .main import BaseSettings, SettingsConfigDict
/usr/lib/python3.9/site-packages/pydantic_settings/main.py:7: in <module>
    from pydantic._internal._config import config_keys
E   ModuleNotFoundError: No module named 'pydantic._internal'
======================================================================================================================================================== short test summary info =========================================================================================================================================================
ERROR pccommon/tests/test_timeouts.py
ERROR pccommon/tests/test_tracing.py
ERROR pccommon/tests/config/test_render_config.py
```

I cannot replicate outside the docker env, this might be an issue when compiling pydantic rust code 🤷

File: pccommon/pccommon/config/core.py [hunk 8]
```diff
@@ -46,6 +46,15 @@ class PCAPIsConfig(BaseSettings):
 
     debug: bool = False
 
+    model_config = {
+        "env_prefix": ENV_VAR_PCAPIS_PREFIX,
+        "env_nested_delimiter": "__",
+        # Mypi is complaining about this with
+        # error: Incompatible types (expression has type "str",
+        # TypedDict item "extra" has type "Extra")
+        "extra": "ignore",  # type: ignore
```
Comment (vincentsarago) [inline pccommon/pccommon/config/core.py:55]: todo: remove `type: ignore` and see if this pass in the CI

File: pccommon/tests/config/test_render_config.py [hunk 9]
```diff
@@ -74,12 +74,12 @@ def test_get_render_config() -> None:
 
 def test_render_config_parse_max_items() -> None:
     config = {
-        "render_params": [],
+        "render_params": {},
```
Comment (vincentsarago) [inline pccommon/tests/config/test_render_config.py:77]: not sure how this passed before

File: pcfuncs/Dockerfile [hunk 10]
```diff
@@ -1,4 +1,4 @@
-FROM mcr.microsoft.com/azure-functions/python:4-python3.8
+FROM mcr.microsoft.com/azure-functions/python:4-python3.10
```
Comment (vincentsarago) [inline pcfuncs/Dockerfile:1]: can revert to 3.8 if needed

File: pcfuncs/animation/settings.py [hunk 11]
```diff
@@ -18,9 +18,10 @@ class AnimationSettings(BaseExporterSettings):
     output_storage_url: str = DEFAULT_ANIMATION_CONTAINER_URL
     tile_request_concurrency: int = DEFAULT_CONCURRENCY
 
-    class Config:
-        env_prefix = ANIMATION_SETTINGS_PREFIX
-        env_nested_delimiter = "__"
+    model_config = {
+        "env_prefix": ANIMATION_SETTINGS_PREFIX,
+        "env_nested_delimiter": "__",  # type: ignore
```
Comment (vincentsarago) [inline pcfuncs/animation/settings.py:23]: same issue locally with mypy complaining about `error: Incompatible types (expression has type "str",`

File: pcfuncs/image/settings.py [hunk 12]
```diff
@@ -15,42 +11,31 @@
...
+    #     logger.info(f"Concurrency limit: {settings.tile_request_concurrency}")
+    #     return settings
+
+
+@cached(LRUCache(maxsize=100))  # type: ignore
+def get_settings() -> ImageSettings:
+    settings = ImageSettings()  # type: ignore
+    logger.info(f"API URL: {settings.api_root_url}")
+    logger.info(f"Concurrency limit: {settings.tile_request_concurrency}")
+    return settings
```
Comment (vincentsarago) [inline pcfuncs/image/settings.py:41]: replaced `ImageSettings.get` with an external function which should gives the same result.

FYI: Pydantic was complaining about the `.get` method

File: pcstac/tests/resources/test_item.py [hunk 13]
```diff
@@ -153,6 +162,30 @@ async def test_item_search_temporal_window_post(app_client):
...
+
+    # NOTE: YYYY-MM-DD format is not supported anymore
+    params = {
+        "collections": first_item["collection"],
+        "bbox": ",".join([str(coord) for coord in first_item["bbox"]]),
+        "datetime": f"{item_date_before.strftime('%Y-%m-%d')}/"
+        f"{item_date_after.strftime('%Y-%m-%d')}",
+    }
+    resp = await app_client.post("/search", json=params)
+    assert resp.status_code == 400
```
Comment (vincentsarago) [inline pcstac/tests/resources/test_item.py:186]: ⚠️ in previous version it seems that `YYYY-MM-DD` was supported, this is not supported anymore.

we could adapt the Search model to relaxe the validation to accept the `old` format 

cc @mmcfarland

Comment (mmcfarland) [reply pcstac/tests/resources/test_item.py:186]: I think that'd be best. Reviewing the example notebooks we render on the site, we do demonstrate using `YYYY-MM-DD`

Comment (vincentsarago) [reply pcstac/tests/resources/test_item.py]: resolved in https://github.com/microsoft/planetary-computer-apis/pull/222/commits/262dc53167abb66469243eb5c9eff68fb994e6d1

**Comment by vincentsarago** [comment]
@mmcfarland I'll all done with the dependency updates! 

Sadly the `pctiler` app is not really well tested so the CI won't show if there are breaking changes 😬 

I'm not quite aware of the interaction pattern between the Client and the API so I'm just going to 🤞 😅

**Comment by mmcfarland** [comment]
Great! I'll pull this down and do some testing and let you know how things are looking!

**Comment by vincentsarago** [comment]
FYI: there are some breaking changes that needs to happen in stac-fastapi (ref: https://github.com/stac-utils/stac-fastapi/issues/713) so we might need to update this PR

File: pcstac/pcstac/main.py [hunk 14]
```diff
@@ -50,6 +55,15 @@
 
 app_settings = get_settings()
 
+items_get_request_model = PCItemCollectionUri
```
Comment (vincentsarago) [inline pcstac/pcstac/main.py:58]: we can now directly define the `GET - /items` model. This wasn't used previously

**Comment by mmcfarland** [comment]
Superseded by #233 -- thanks for this baseline @vincentsarago!

**Pull request closed without merging**